### PR TITLE
fix(consensus): resend catch-up block parts to prevent lagging-peer wedge

### PR DESCRIPTION
**Pull request by lklimek**
## Why this PR exists

**Problem.** Consensus catch-up block-part gossip optimistically marks a part as *delivered* the moment it is sent (`updatePeerProposalBlockPart`). A peer that is catching up can legitimately be at a step where it is not yet expecting block parts (RoundStepNewHeight/Propose with a nil `ProposalBlockParts`) and silently drops the part it receives (`state_add_prop_block.go`, "received a block part when we are not expecting any"). Because the sender now believes the peer has the part, `GossipBlockPartsForCatchup`'s `prs.ProposalBlockParts.Not().PickRandom()` finds nothing missing and **never resends it**. When the catch-up *commit* then arrives ahead of the proposal, the node stores the commit and builds an empty part set from the commit header (`state_data.go`, "Commit came in before proposal") and waits forever for a part that is never resent — the node **wedges permanently** at that height.

**What breaks without it.** Under timing pressure a lagging node freezes (observed at height ~17 and ~19 in repro) while the rest of the network advances. This surfaces as the flaky CI test `TestReactorValidatorSetChanges`, which intermittently times out at 120s. In production this normally self-heals via propose/round timeouts (which reset peer round state and re-trigger gossip) and via blocksync; the consensus-only test harness — with its one-shot mock ticker (no round timeouts) and no blocksync — disables both recovery paths and so exposes the latent bug as a hard hang.

## What was done

- `internal/consensus/gossiper.go`: the **catch-up** block-part path no longer optimistically records the part as delivered, so it keeps resending the missing part each gossip tick until the peer applies the block and advances its height. This mirrors the "resend until acked" behavior already used for catch-up commits and removes catch-up's dependency on timeouts to recover a dropped part. `syncProposalBlockPart` gained a `markPeerHasPart bool`; **regular same-height block-part gossip is unchanged** (still marks optimistically, for efficiency).
- Added regression test `TestGossipBlockPartsForCatchupResends` asserting catch-up does not mark the peer as having the part and resends on the next tick.

## Testing

- Repro: 2 permanent wedges captured in a 12-way `-race` stress sample (nodes frozen at heights 17 and 19, with the "Commit came in before proposal" signature after ~100 blocks produced).
- After fix: **0 wedges** across ~600 `-race` runs (single-instance 25/25, 6-way 144/144, plus 10–12-way runs) and 600 plain runs. The only residual failures under heavy parallelism were CPU-starvation artifacts (whole instance produces 0 blocks) from oversubscribing the host — identical for baseline and fixed, and unrelated to this change.
- New deterministic unit test `TestGossipBlockPartsForCatchupResends` passes; full `TestGossiper` suite passes.
- `gofmt` / `go vet` / `golangci-lint` (changed files) clean.
- **Independently verified** by an adversarial QA review: root-cause, fix-correctness, and test-validity all confirmed, with 0 blocking concerns (regression risk assessed LOW — bounded resends, no gossip storm, guaranteed termination once the peer advances).

## Breaking changes

None.

## Known tradeoff / follow-up

For multi-part blocks, catch-up may redundantly resend parts (there is no per-part dedup/ack on the catch-up path). This is bounded — O(gossip_ticks × parts × peers) until the lagging peer executes the block and advances — and is fine for a fallback path (blocksync covers large gaps). It is strictly better than a permanent wedge. A per-part acknowledgement is a possible future optimization, not a correctness issue.

## Scope

This fixes **one** source of shard-01 CI flakiness — the consensus reactor valset-change wedge. Other shard-01 flakes are separate and not addressed here (e.g. `TestHandshakeReplayNone` `t.TempDir()` teardown race; a still-open evidence-reactor propagation flake).

<sub>🤖 Co-authored by [Claudius the Magnificent](https://github.com/lklimek/claudius) AI Agent</sub>

### CODE REVIEW

**Comment by lklimek** [comment]
@thepastaclaw please review

**Comment by lklimek** [comment]
@thepastaclaw please review

<sub>🤖 Co-authored by [Claudius the Magnificent](https://github.com/lklimek/claudius) AI Agent</sub>

**Comment by thepastaclaw** [comment]
✅ Review complete (commit 042e7b5d)

File: internal/consensus/gossiper.go [hunk 1]
```diff
@@ -192,7 +195,18 @@ func (g *msgGossiper) GossipBlockPartsForCatchup(
...
+	//
+	// A lagging peer can be at a step where it is not yet expecting block parts
+	// (e.g. RoundStepNewHeight/Propose with a nil ProposalBlockParts) and will
+	// silently drop the part. If we optimistically marked it delivered, our view
+	// of the peer would show the full part set and we would never resend it,
+	// leaving the peer wedged with a stored commit but an incomplete block (it
+	// learns the part-set header only once the catch-up commit arrives). By not
+	// marking it, we keep resending the missing part each gossip tick until the
+	// peer actually applies the block and advances its height, mirroring the
+	// "resend until acked" behavior already used for catch-up commits.
```
Comment (thepastaclaw) [inline internal/consensus/gossiper.go]: **💬 Nitpick: Comment claims a non-existent "resend until acked" pattern for catch-up commits**

The new comment ends with "mirroring the 'resend until acked' behavior already used for catch-up commits." That behavior does not exist: `GossipCommit` (gossiper.go:215-246) early-returns when `prs.HasCommit` is set, and `updatePeerCommit` -> `SetHasCommit` flips that flag the first time the commit is sent. Commits are sent once optimistically and then suppressed, not resent until acked. The catch-up part resend is actually the only path in this file that keeps retrying, and documenting a false invariant in consensus-critical gossip code risks future regressions where someone removes the catch-up resend on the assumption that commits will pick up the slack.

```suggestion
	// marking it, we keep resending the missing part each gossip tick until the
	// peer actually applies the block and advances its height.
```

<sub>source: ['codex']</sub>

File: internal/consensus/gossiper_test.go [hunk 2]
```diff
@@ -404,6 +404,50 @@ func (suite *GossiperSuiteTest) TestGossipBlockPartsForCatchup() {
...
+		Return(nil)
+
+	// First tick: send the missing part.
+	suite.gossiper.GossipBlockPartsForCatchup(ctx, cstypes.RoundState{}, suite.ps.GetRoundState())
+	suite.Require().False(suite.ps.PRS.ProposalBlockParts.GetIndex(0),
+		"catch-up must not optimistically mark the peer as having the part")
+
+	// Second tick: the part is still considered missing, so it is resent.
+	suite.gossiper.GossipBlockPartsForCatchup(ctx, cstypes.RoundState{}, suite.ps.GetRoundState())
+}
```
Comment (thepastaclaw) [inline internal/consensus/gossiper_test.go:450]: **🟡 Suggestion: Regression test only exercises a single-part block**

`TestGossipBlockPartsForCatchupResends` asserts `partSet.Total() == 1` (line 418), so `PickRandom()` over `BitArray().Not()` always returns index 0. The test cannot distinguish "catch-up correctly iterates missing indices" from "the only missing index gets re-picked" — it locks in the single-part wedge case (which is exactly the failure mode the PR fixes) but doesn't document behavior on realistic multi-part blocks where the PR description explicitly notes redundant sends without per-part dedup. A multi-part variant (e.g. `Total() >= 3`) asserting that after N ticks every index has been Sent at least once would lock in the broader resend invariant.

<sub>source: ['claude']</sub>

**Review by thepastaclaw — COMMENTED** [review]
## Code Review

Targeted, intentional fix for a consensus catch-up wedge: removes the optimistic mark on catch-up block parts so a lagging peer keeps receiving them until it actually advances. The change is well-scoped (sender-only, no signing/encoding/validation surface) and ships with a clear regression test. Findings are limited to two misleading new/stale comments, a thin single-part test, and one explicitly acknowledged tradeoff (unbounded resend against a non-progressing peer) that both agents flagged.

🟡 2 suggestion(s) | 💬 2 nitpick(s)

_2 additional finding(s) omitted (not in diff)._

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `internal/consensus/gossiper.go`:
- [SUGGESTION] internal/consensus/gossiper.go:169-213: Catch-up resend has no per-peer rate floor or eviction path
  With `markPeerHasPart=false`, the local view of the peer's `ProposalBlockParts` bitarray never converges as long as the peer reports a stale height, so every `PeerGossipSleepDuration` tick (100ms default, 5ms validator profile) we call `loadMeta` + `loadPart` (disk reads) and push another block-part envelope. A peer slot occupied by a non-progressing node — accidentally stuck or adversarial — keeps the path hot indefinitely; pre-PR a single pass over the part set put a natural ceiling on this. The PR description already calls this out as an acknowledged tradeoff and the fix is strictly safer than the prior wedge (which was a chain-liveness loss, not just bandwidth), so this is not blocking — but a follow-up to add a coarse per-peer/per-part cooldown, or to bound retries before reconsidering the peer, would close the amplification gap without re-introducing the wedge.

In `internal/consensus/gossiper_test.go`:
- [SUGGESTION] internal/consensus/gossiper_test.go:413-449: Regression test only exercises a single-part block
  `TestGossipBlockPartsForCatchupResends` asserts `partSet.Total() == 1` (line 418), so `PickRandom()` over `BitArray().Not()` always returns index 0. The test cannot distinguish "catch-up correctly iterates missing indices" from "the only missing index gets re-picked" — it locks in the single-part wedge case (which is exactly the failure mode the PR fixes) but doesn't document behavior on realistic multi-part blocks where the PR description explicitly notes redundant sends without per-part dedup. A multi-part variant (e.g. `Total() >= 3`) asserting that after N ticks every index has been Sent at least once would lock in the broader resend invariant.
```
</details>

**Review by thepastaclaw — APPROVED** [review]
## Code Review

Narrow, consensus-safe liveness fix for catch-up block-part gossip. Catch-up sends no longer optimistically mark parts as delivered, so a lagging peer at RoundStepNewHeight/Propose that drops the part keeps receiving it until it advances. The same-height gossip fast path is preserved via the new markPeerHasPart flag, the corrected comment removes a misleading 'resend until acked' claim, and the added multi-part regression test exercises iteration across all missing indices.

**Review by thepastaclaw — APPROVED** [review]
## Code Review

PR #1365 is a narrowly scoped, correct fix for a catch-up gossip wedge in internal/consensus/gossiper.go. Same-height gossip behavior is preserved while catch-up sends no longer optimistically mark parts as delivered, ensuring resends continue until the peer applies the block. The sync() helper safely handles a nil syncFunc, and the change ships with both a deterministic regression test and a multi-part resend test. All four agent reviews converged on no in-scope findings.

**Comment by lklimek** [comment]
@thepastaclaw please re-review — your two prior suggestions are addressed:
- `ff7d711`: corrected the false "resend until acked" comment (verified `GossipCommit` is one-shot optimistic) and added `TestGossipBlockPartsForCatchupResendsMultiPart` (3-part block, coupon-collector non-flaky) locking in the multi-part resend invariant;
- `042e7b5`: fixed a US-locale `misspell` lint blocker (`serialises`→`serializes`).

CI should now be fully green (all 6 shards were already passing; misspell was the only blocker).

<sub>🤖 Co-authored by [Claudius the Magnificent](https://github.com/lklimek/claudius) AI Agent</sub>
